### PR TITLE
DTSPO-15778 migrate Backstage to App Proxy - part one

### PR DESCRIPTION
**Pull request by charlesluokb**
### Jira link (if applicable)

https://tools.hmcts.net/jira/browse/DTSPO-15778

### Change description ###
The ticket is divided into two parts/tasks

__Part 1__: Grant delegated permissions (e.g. OpenID Connect scopes `email` , `profile` , `offline_access`, and `openid`) programatically. 

__Part 2__: Migrate Backstage onto the Application Proxy. This PR (https://github.com/hmcts/azure-app-proxy/pull/40) has partially addressed it but requires code changes from Part 1 to grant delegated permissions. 

This PR is mainly for Part 1

I have added a comment in the ticket (https://tools.hmcts.net/jira/browse/DTSPO-15778) with more details 

### Checklist

- [x] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/applicationManager.test.ts [hunk 1]
```diff
@@ -157,11 +158,6 @@ describe("applicationManager", () => {
       optionalClaims: [{ name: "groups", additionalProperties: [] }],
     });
 
-    let application = await readApplication({
```
Comment (charlesluokb) [inline src/applicationManager.test.ts:160]: the application  is re-retrived at line 176 and then referenced afterwards hence I removed this call

File: src/applicationManager.ts [hunk 2]
```diff
@@ -513,9 +516,85 @@ export async function setResourceAccess({
       },
     );
     await errorHandler("Granting graph api permissions", assignRolesResult);
+  } else {
+    console.log("No graph api permissions to grant");
+  }
+}
+
+function _buildResourceAccess(roleIds: string[], accessType: string) {
```
Comment (timja) [inline src/applicationManager.ts:524]: no need to prefix with `_` if you don't export the function it won't be available

File: src/applicationManager.ts [hunk 3]
```diff
@@ -513,9 +516,85 @@ export async function setResourceAccess({
...
+}
+
+async function _buildRequiredResourceAccessPayload(currRequiredResourceAccessList: any[], graphAppId: string, appRoleIds: string[],
+                                                   scopeIds: string[]) {
+
+  /* todo: to remove current resource access and then build a new one based on access roles and access scopes. However we need to deal with
+           the eventual consistency of the graph API. */
+  let roleAccess = _buildResourceAccess(appRoleIds, "Role") ?? []
+  let scopeAccess = _buildResourceAccess(scopeIds, "Scope") ?? []
+  console.log("currRequiredResourceAccessObj", JSON.stringify(currRequiredResourceAccessList))
```
Comment (timja) [inline src/applicationManager.ts:538]: is this still needed?

File: src/applicationManager.ts [hunk 4]
```diff
@@ -513,9 +516,85 @@ export async function setResourceAccess({
...
+      headers: {
+        Authorization: `Bearer ${token}`,
+        "Content-Type": "application/json",
+      },
+    },
+  );
+  await errorHandler("Getting MS Graph oAuth2 permission scopes", graphResponse);
+
+  const graphAppObj = (await graphResponse.json()).value[0];
+  // const graphAppObjId = graphAppObj.id;
```
Comment (timja) [inline src/applicationManager.ts:575]: is this needed?

File: src/main.ts [hunk 5]
```diff
@@ -44,7 +44,7 @@ const credential = new DefaultAzureCredential();
 const { token } = await credential.getToken(
   "https://graph.microsoft.com/.default",
 );
-
+console.log("Got token", token);
```
Comment (timja) [inline src/main.ts:47]: 🚫 remove

File: src/applicationManager.ts (hunk 2 above)
Comment (charlesluokb) [reply src/applicationManager.ts:524]: prefix with `_` purely for readability. I know it is a private function when seeing it 😃

File: src/applicationManager.ts (hunk 3 above)
Comment (charlesluokb) [reply src/applicationManager.ts:538]: I would keep it as it is useful in the log

File: src/applicationManager.ts (hunk 4 above)
Comment (charlesluokb) [reply src/applicationManager.ts:575]: well spotted! it is not needed

File: src/main.ts (hunk 5 above)
Comment (charlesluokb) [reply src/main.ts:47]: sure thing :-)

**Review by timja — APPROVED** [review]

**Pull request merged**